src/cmds/hardware/pin/pin.c

```diff
@@ -10,71 +10,126 @@
-	printf("USAGE:\npin [pin name] [state]\n");
+	printf("USAGE:\npin [pin name] [pin index] [state]\n");
+	// TODO: add examples
 }
 
-static void *gpio_by_name(char *name) {
-	if (!strncmp("GPIO", name, 4) || !strncmp("gpio", name, 4))
-		name += 4;
+// See, for example, stm32f4xx_hal_gpio.h. Single pin maps to single bit.
+// 15 pins are set means 0xFFFF. No pins set means 0x0000.
+#define PIN_UNSPECIFIED (0)
 
-	if (name[0] == 'a' || name[0] == 'A')
-		return GPIOA;
-	else if (name[0] == 'b' || name[0] == 'B')
-		return GPIOB;
-	else if (name[0] == 'c' || name[0] == 'C')
-		return GPIOC;
-	else if (name[0] == 'd' || name[0] == 'D')
-		return GPIOD;
-	else if (name[0] == 'e' || name[0] == 'E')
-		return GPIOE;
-	else if (name[0] == 'f' || name[0] == 'F')
-		return GPIOF;
-	
-	return NULL;
+/**
+ * @brief [pin_index_of_int n] converts index [n] of pin to concrete GPIO_PIN_[n].
+ *        Returns 0 on fail.
+ */
+static uint16_t pin_index_of_int(int n) {
+	switch(n) {
+	case 0:		return GPIO_PIN_0;
+	case 1:		return GPIO_PIN_1;
+	case 2:		return GPIO_PIN_2;
+	case 3:		return GPIO_PIN_3;
+	case 4:		return GPIO_PIN_4;
+	case 5:		return GPIO_PIN_5;
+	case 6:		return GPIO_PIN_6;
+	case 7:		return GPIO_PIN_7;
+	case 8:		return GPIO_PIN_8;
+	case 9:		return GPIO_PIN_9;
+	case 10:	return GPIO_PIN_10;
+	case 11:	return GPIO_PIN_11;
+	case 12:	return GPIO_PIN_12;
+	case 13:	return GPIO_PIN_13;
+	case 14:	return GPIO_PIN_14;
+	case 15:	return GPIO_PIN_15;
+	default:	return 0;
+	}
 }
 
 static uint16_t pin_by_name(char *name) {
-	int n = atoi(name);
-	switch(n) {
-	case 0:
-		return GPIO_PIN_0;
-	case 1:
-		return GPIO_PIN_1;
-	case 2:
-		return GPIO_PIN_2;
-	case 3:
-		return GPIO_PIN_3;
-	case 4:
-		return GPIO_PIN_4;
-	case 5:
-		return GPIO_PIN_5;
-	case 6:
-		return GPIO_PIN_6;
-	case 7:
-		return GPIO_PIN_7;
-	case 8:
-		return GPIO_PIN_8;
-	case 9:
-		return GPIO_PIN_9;
-	case 10:
-		return GPIO_PIN_10;
-	case 11:
-		return GPIO_PIN_11;
-	case 12:
-		return GPIO_PIN_12;
-	case 13:
-		return GPIO_PIN_13;
-	case 14:
-		return GPIO_PIN_14;
-	case 15:
-		return GPIO_PIN_15;
-	default:
+	char *end;
+	int n = strtol(name, &end, 10);
+	if (*end == '\0')
+		return pin_index_of_int(n);
+	else {
+		fprintf(stderr, "Can't extract pin index from string '%s'\n", name);
+		return PIN_UNSPECIFIED;
+	}
+}
+
+/**
+ * @brief gpio_by_name Parses GPIO and (maybe) pin index from input string.
+ * @param name Input string to parse
+ * @param gpio Output parameter. If function returns successfully contains
+ *             information about specified GPIO.
+ * @param pin  Output parameter. If pin is specified contains index of specified
+ *             pin. If pin is not specified contains PIN_UNSPECIFIED.
+ *             If function fails result value is not specified.
+ * @return TRUE if functions successfully parsed GPIO and (maybe) pin index
+ *         from input string. On fail returns FALSE.
+ */
+static int gpio_by_name(char *name, void **gpio, gpio_mask_t *pin) {
```

Comment:
I think it's a good idea to return pointer to GPIO address and get rid of **gpio argument, this way code will be shorter and more understandable.